tests/baselines/reference/genericRestParameters3.errors.txt

```diff
@@ -69,16 +70,17 @@ tests/cases/conformance/types/rest/genericRestParameters3.ts(59,5): error TS2345
     ~~
 !!! error TS2322: Type '(x: string, y: string) => void' is not assignable to type '(x: string, ...args: [string] | [number, boolean]) => void'.
 !!! error TS2322:   Types of parameters 'y' and 'args' are incompatible.
-!!! error TS2322:     Type '[string] | [number, boolean]' is not assignable to type '[y: string]'.
-!!! error TS2322:       Type '[number, boolean]' is not assignable to type '[y: string]'.
-!!! error TS2322:         Source has 2 element(s) but target allows only 1.
+!!! error TS2322:     Type '[string] | [number]' is not assignable to type '[y: string]'.
+!!! error TS2322:       Type '[number]' is not assignable to type '[y: string]'.
+!!! error TS2322:         Type 'number' is not assignable to type 'string'.
     f1 = f3;  // Error
     ~~
 !!! error TS2322: Type '(x: string, y: number, z: boolean) => void' is not assignable to type '(x: string, ...args: [string] | [number, boolean]) => void'.
 !!! error TS2322:   Types of parameters 'y' and 'args' are incompatible.
-!!! error TS2322:     Type '[string] | [number, boolean]' is not assignable to type '[y: number, z: boolean]'.
-!!! error TS2322:       Type '[string]' is not assignable to type '[y: number, z: boolean]'.
-!!! error TS2322:         Source has 1 element(s) but target requires 2.
+!!! error TS2322:     Type '[number, boolean] | [string, undefined]' is not assignable to type '[y: number, z: boolean]'.
+!!! error TS2322:       Type '[string, undefined]' is not assignable to type '[y: number, z: boolean]'.
+!!! error TS2322:         Type at position 0 in source is not compatible with type at position 0 in target.
+!!! error TS2322:           Type 'string' is not assignable to type 'number'.

```

Comment:
This error change comes from the fact that I fill the slice of the target type with undefineds here (when it's shorter than the source's slice):
https://github.dev/microsoft/TypeScript/blob/6c5383d701417e8a98398b4171af0f201ee593c5/src/compiler/checker.ts#L19394-L19402

I think this is a good change. This will allow more patterns to be assignable, especially with target signatures using unions of tuples for their rest parameter. After all, the implementation is free to ignore any of the provided arguments etc. This is also already just fine in TS:
```ts
declare let f1: (a: string) => void
declare let f2: (a: string, b: number) => void

f2 = f1
```

And since this is OK I think that this one should be too:
```ts
declare let f1: (x: string, ...args: [string] | [number, boolean]) => void;

// Type '(a: string, b: string | number, c: boolean | undefined) => void' is not assignable to type '(x: string, ...args: [string] | [number, boolean]) => void'.
//   Types of parameters 'b' and 'args' are incompatible.
//     Type '[string] | [number, boolean]' is not assignable to type '[b: string | number, c: boolean | undefined]'.
//       Type '[string]' is not assignable to type '[b: string | number, c: boolean | undefined]'.
//         Source has 1 element(s) but target requires 2.
f1 = (a, b, c) => {}
```

So with this change that caused this particular error here to be reported differently we now allow the example above and I added a test case for it:
https://github.com/microsoft/TypeScript/blob/46f2e57d76f26b74f596f2acb5d04a30572f13a7/tests/cases/compiler/restTupleUnionWithRestContextualParams.ts#L6